src/tests/ecdh.rs

```diff
@@ -0,0 +1,154 @@
+    let ret = fn_find_objects_init(session, template.as_ptr() as *mut _, 1);
+    assert_eq!(ret, CKR_OK);
+    let mut count: CK_ULONG = 0;
+    let ret = fn_find_objects(session, &mut handle, 1, &mut count);
+    assert_eq!(ret, CKR_OK);
+    assert_eq!(count, 1);
+    assert_ne!(handle, CK_INVALID_HANDLE);
+    let ret = fn_find_objects_final(session);
+    assert_eq!(ret, CKR_OK);
+
+    /* Peer key */
+    let _peer_value = hex::decode(
+        "d40b07b1ea7b86d4709ef9dc634c61229feb71abd63dc7fc85ef46711a87b2\
+         10",
+    )
+    .expect("Failed to decode peer value");
+    let mut peer_point = hex::decode(
+        "04fbcea7c2827e0e8085d7707b23a3728823ea6f4878b24747fb4fd2842d40\
+         6c732393c85f1f710c5afc115a39ba7e18abe03f19c9d4bb3d47d19468b818\
+         efa535",
+    )
+    .expect("Failed to decode peer point");
+
+    /* derive key without shared data */
+    let mut params = CK_ECDH1_DERIVE_PARAMS {
+        kdf: CKD_NULL,
+        ulSharedDataLen: 0,
+        pSharedData: std::ptr::null_mut(),
+        ulPublicDataLen: peer_point.len() as CK_ULONG,
+        pPublicData: peer_point.as_mut_ptr(),
+    };
+    let mut mechanism: CK_MECHANISM = CK_MECHANISM {
+        mechanism: CKM_ECDH1_DERIVE,
+        pParameter: &mut params as *mut _ as CK_VOID_PTR,
+        ulParameterLen: std::mem::size_of::<CK_ECDH1_DERIVE_PARAMS>()
+            as CK_ULONG,
+    };
+
+    let derive_template = make_attr_template(
+        &[
+            (CKA_CLASS, CKO_SECRET_KEY),
+            (CKA_KEY_TYPE, CKK_AES),
+            (CKA_VALUE_LEN, 16),
+        ],
+        &[],
+        &[(CKA_ENCRYPT, true), (CKA_DECRYPT, true)],
+    );
+
+    let mut s_handle = CK_INVALID_HANDLE;
+    let ret = fn_derive_key(
+        session,
+        &mut mechanism,
+        handle,
+        derive_template.as_ptr() as *mut _,
+        derive_template.len() as CK_ULONG,
+        &mut s_handle,
+    );
+    assert_eq!(ret, CKR_OK);
+
+    /* Test the derived key works */
+    /* Data need to be exactly one block in size for CBC */
+    let data = "0123456789ABCDEF";
+    let iv = "FEDCBA0987654321";
+    let enc = ret_or_panic!(encrypt(
+        session,
+        s_handle,
+        data.as_bytes(),
+        &CK_MECHANISM {
+            mechanism: CKM_AES_CBC,
+            pParameter: void_ptr!(iv.as_bytes()),
+            ulParameterLen: iv.len() as CK_ULONG,
+        }
+    ));
+    assert_eq!(enc.len(), 16);
+
+    let dec = ret_or_panic!(decrypt(
+        session,
+        s_handle,
+        enc.as_slice(),
+        &CK_MECHANISM {
+            mechanism: CKM_AES_CBC,
+            pParameter: void_ptr!(iv.as_bytes()),
+            ulParameterLen: iv.len() as CK_ULONG,
+        }
+    ));
+    assert_eq!(dec.len(), data.len());
+    assert_eq!(data.as_bytes(), dec.as_slice());
+
+    /* derive with shared data (and cofactor) */
+    let shared = "shared data";
+    let mut params = CK_ECDH1_DERIVE_PARAMS {
+        kdf: CKD_SHA512_KDF,
```

Comment:
Can we add an explicit test with the SP800 variants?
You may look at how I did testing in tests/kdfs.rs where I have an array of values and run through the testing multiple times so many combinations are tested.

Reply:
Added test for SP800 variant. I thought I could use the test vectors for that, but from their description I was not able to figure out how to plug them together to get the right answer. The parsing and calling code is there, but the resulting key is not checked.